### PR TITLE
feat(storage): release download handles sooner

### LINKED ISSUE #6160: Release CURL handle as soon as the download completes
In a `CurlDownloadRequest` the curl handle is released in the destructor:

https://github.com/googleapis/google-cloud-cpp/blob/9ebc00b04b4fd2767c5bee6acdd1172407e6b9eb/google/cloud/storage/internal/curl_download_request.h#L56-L57

They can be released earlier, i.e., once the download completes.  That makes the resources usable to other streams sooner.

### PR DESCRIPTION
**Pull request by coryan**
Release the CURL* handles as soon as we are done using them. The
downloads used to hold on to these until the `CurlDownloadRequest`
object was deleted, which in turn could live until the
`storage::ObjectReadStream` was deleted.

Fixes #6160

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/7085)

### CODE REVIEW

File: google/cloud/storage/tests/curl_download_request_integration_test.cc [hunk 1]
```diff
@@ -75,6 +77,134 @@ TEST(CurlDownloadRequestTest, SimpleStream) {
...
+  Status status;
+  for (int i = 0; i != 3; ++i) {
+    status = download();
+    if (status.ok()) break;
+    std::this_thread::sleep_for(delay);
+    delay *= 2;
+  }
+  EXPECT_EQ(1, factory->CurrentHandleCount());
+  EXPECT_EQ(1, factory->CurrentMultiHandleCount());
+  ASSERT_STATUS_OK(status);
```
Comment (devbww) [inline google/cloud/storage/tests/curl_download_request_integration_test.cc:120]: Should this happen before the handle-count expectations?

File: google/cloud/storage/tests/curl_download_request_integration_test.cc [hunk 2]
```diff
@@ -75,6 +77,134 @@ TEST(CurlDownloadRequestTest, SimpleStream) {
...
+
+TEST(CurlDownloadRequestTest, SimpleStreamReadAfterClosed) {
+  auto constexpr kLineCount = 10;
+  auto download = [&]() -> StatusOr<std::string> {
+    std::string contents;
+    CurlRequestBuilder builder(
+        HttpBinEndpoint() + "/stream/" + std::to_string(kLineCount),
+        storage::internal::GetDefaultCurlHandleFactory());
+    auto download = std::move(builder).BuildDownloadRequest();
+    // Perform a series of very small `.Read()` calls. This will force the
```
Comment (devbww) [inline google/cloud/storage/tests/curl_download_request_integration_test.cc:176]: "This will force the ...."?

**Review by devbww — APPROVED** [review]

File: google/cloud/storage/tests/curl_download_request_integration_test.cc (hunk 1 above)
Comment (coryan) [reply google/cloud/storage/tests/curl_download_request_integration_test.cc:121]: Added a comment. It should work even if `ASSERT_STATUS_OK()` fails.  Well, at least for now, I have another PR that will change this.

File: google/cloud/storage/tests/curl_download_request_integration_test.cc (hunk 2 above)
Comment (coryan) [reply google/cloud/storage/tests/curl_download_request_integration_test.cc]: Fixed, thanks.

**Pull request merged**